### PR TITLE
get datums using pallas observer

### PR DESCRIPTION
**Pull request by falcucci**
## Content

Includes:
- get current datums using the new pallas observer;
- brings all compatibilities with the previous cardano-cli format;
- uses the current cardano-cli as a default fallback for the remaining methods;

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [ ] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested

## Comments

The automatic fallback mode has been implemented to make sure we progressively rollout the new integration.

### CODE REVIEW

**Comment by falcucci** [comment]
@scarmuega ready for review,

we also need a new pallas release to bump the new version with the get utxo by address state query in order to fix the [check](https://github.com/txpipe/mithril/actions/runs/7290537684/job/19867554439?pr=1) ci step :)

**Comment by falcucci** [comment]
@scarmuega the canonical trait worked like a charm!

all the requests was addressed and is ready for a new review.

thanks

File: mithril-common/src/chain_observer/codec.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    }
+}
+
+/// Inspects the given bytes and returns a decoded `R` instance.
+pub fn inspect<R>(inner: Vec<u8>) -> R
+where
+    for<'b> R: pallas_codec::minicbor::Decode<'b, ()>,
+{
+    minicbor::decode(&inner).unwrap()
+}
```
Comment (scarmuega) [inline mithril-common/src/chain_observer/codec.rs]: This is just short boilerplate code, no need to put it on a different mod. I would inline it inside cli_observer.

File: mithril-common/src/chain_observer/pallas_observer.rs [hunk 2]
```diff
@@ -78,6 +84,81 @@ impl PallasChainObserver {
...
+    /// Fetches the current UTxO by address using the provided `statequery` client.
+    async fn get_utxo_by_address(
+        &self,
+        statequery: &mut Client,
+        address: &ChainAddress,
+    ) -> StdResult<UTxOByAddress> {
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| anyhow!(err))
```
Comment (scarmuega) [inline mithril-common/src/chain_observer/pallas_observer.rs:156]: is `map_err(|err| anyhow!(err))` necessary? doesn't anyhow already transform the error type automatically?

**Review by scarmuega — APPROVED** [review]

File: mithril-common/src/chain_observer/pallas_observer.rs (hunk 2 above)
Comment (falcucci) [reply mithril-common/src/chain_observer/pallas_observer.rs:156]: yes but on mithril they use as much as possible `with_context` to provide additional context in the error type, so due to that the `map_err` should be needed AFAIK

File: mithril-common/src/chain_observer/model.rs [hunk 3]
```diff
@@ -1,20 +1,49 @@
...
 use strum::{Display, EnumDiscriminants};
 use thiserror::Error;
 
 use crate::{StdError, StdResult};
 
-/// [Datum] represents an inline datum from UTxO.
-#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
-#[serde(rename_all = "lowercase")]
-pub struct Datum(pub PlutusData);
+cfg_fs! {
```
Comment (scarmuega) [inline mithril-common/src/chain_observer/model.rs:10]: Why is this macro require? does it have to do with fs = filesystem? I don't see any access to the file system in this block.

File: mithril-common/src/chain_observer/pallas_observer.rs [hunk 4]
```diff
@@ -69,51 +68,66 @@ impl PallasChainObserver {
             .acquire(None)
             .await
             .map_err(|err| anyhow!(err))
-            .with_context(|| "PallasChainObserver Failed to acquire statequery")?;
+            .with_context(|| "PallasChainObserver failed to acquire statequery")?;
```
Comment (scarmuega) [inline mithril-common/src/chain_observer/pallas_observer.rs:77]: I think that for static message errors is more efficient to use `context(str)` instead of `with_context(fn)`

File: mithril-common/src/lib.rs [hunk 5]
```diff
@@ -10,6 +10,16 @@
...
+macro_rules! cfg_fs {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "fs")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "fs")))]
+            $item
+        )*
+    }
+}
+
```
Comment (scarmuega) [inline mithril-common/src/lib.rs]: why is this required?

Comment (falcucci) [reply mithril-common/src/lib.rs]: due to wasm [build](https://github.com/input-output-hk/mithril/actions/runs/7422372066/job/20197520073?pr=1403) errors, the mithril team suggested a conditional compilation macro.

more details here: https://github.com/input-output-hk/mithril/pull/1403/files/acccd103df4ad21cd126de6e14003135da7105b0#r1442959628

File: mithril-common/src/chain_observer/pallas_observer.rs (hunk 4 above)
Comment (falcucci) [reply mithril-common/src/chain_observer/pallas_observer.rs:77]: the previous implementation was using `context` but it was requested to use `with_context` on non test code.

It seems to be an ADR from the mithril team, here: https://mithril.network/doc/adr/6
